src/libstore/ssh-store.cc

```diff
@@ -89,17 +92,76 @@ class SSHStore : public virtual SSHStoreConfig, public virtual RemoteStore
     };
 };
 
+/**
+ * The mounted ssh store assumes that filesystems on the remote host are shared
+ * with the local host. This means that the remote nix store is available
+ * locally and is therefore treated as a local filesystem store.
+ */
+class MountedSSHStore : public virtual SSHStore, public virtual LocalFSStore
+{
+public:
+
+    MountedSSHStore(const std::string & scheme, const std::string & host, const Params & params)
+        : StoreConfig(params)
+        , RemoteStoreConfig(params)
+        , CommonSSHStoreConfig(params)
+        , SSHStoreConfig(params)
+        , Store(params)
+        , RemoteStore(params)
+        , SSHStore(scheme, host, params)
+        , LocalFSStoreConfig(params)
+        , LocalFSStore(params)
+    {
+        extraRemoteProgramArgs = "--process-ops --allow-perm-roots";
+    }
+
+    static std::set<std::string> uriSchemes()
+    {
+        return {"mounted-ssh-ng"};
+    }
+
+    std::string getUri() override
+    {
+        return *uriSchemes().begin() + "://" + host;
+    }
+
+    void narFromPath(const StorePath & path, Sink & sink) override
+    {
+        return SSHStore::narFromPath(path, sink);
+    }
+
+    ref<FSAccessor> getFSAccessor() override
+    {
+        return SSHStore::getFSAccessor();
+    }
+
+    std::optional<std::string> getBuildLogExact(const StorePath & path) override
+    {
+        return SSHStore::getBuildLogExact(path);
```

Comment:
```suggestion
        return LocalFSStore::narFromPath(path, sink);
    }

    ref<FSAccessor> getFSAccessor() override
    {
        return LocalFSStore::getFSAccessor();
    }

    std::optional<std::string> getBuildLogExact(const StorePath & path) override
    {
        return LocalFSStore::getBuildLogExact(path);
```
Don't we want to take advantage of the file system for these?

Reply:
I didn't make this change because I didn't want to do a big behavioral change, just pure refactors. But I still invite you to do this change and see if it still works for you.

Reply:
Ah, good point. I'll give it a try, but if I remember correctly (not sure&mdash;have to re-verify) some of these operations require a writable FS on the client (to grab some locks?). Unfortunately, for us the filesystem is mounted read-only client-side.

In any case, I'll give it a try and let you know how it goes.

Reply:
Great glad it works! Definitely fine to rebase and squash; I don't really care about being attributed to the docs in the history very much :).